File: Tests/MihomoControllerModelsTests.swift
import XCTest

@testable import MihomoMeter

final class MihomoControllerModelsTests: XCTestCase {
  private let decoder = JSONDecoder()

  func testDecodesVersionFixture() throws {
    let response = try decoder.decode(
      MihomoVersionResponse.self,
      from: FixtureLoader.data(named: "version")
    )

    XCTAssertTrue(response.meta)
    XCTAssertEqual(response.version, "v1.19.0")
  }

  func testDecodesProxyCatalogFixture() throws {
    let response = try decoder.decode(
      MihomoProxiesResponse.self,
      from: FixtureLoader.data(named: "proxies")
    )

    XCTAssertEqual(response.proxies["DIRECT"]?.type, "Direct")
    XCTAssertEqual(response.proxies["Synthetic Proxy"]?.type, "Vmess")
    XCTAssertEqual(response.proxies["Synthetic Group"]?.now, "Synthetic Proxy")
  }

  func testDecodesRuntimeConfiguration() throws {
    let data = Data(
      """
      {
        "mode": "rule",
        "allow-lan": false,
        "ipv6": true,
        "mixed-port": 7890,
        "port": 7891,
        "socks-port": 7892,
        "global-ua": "mihomo-test-agent",
        "find-process-mode": "strict",
        "tun": {
          "enable": true,
          "stack": "system",
          "auto-route": true
        }
      }
      """.utf8
    )

    let response = try decoder.decode(
      MihomoRuntimeConfigurationResponse.self,
      from: data
    )
    let configuration = response.runtimeConfiguration

    XCTAssertEqual(configuration.mode, "rule")
    XCTAssertEqual(configuration.allowsLAN, false)
    XCTAssertEqual(configuration.isIPv6Enabled, true)
    XCTAssertEqual(configuration.mixedPort, 7_890)
    XCTAssertEqual(configuration.httpPort, 7_891)
    XCTAssertEqual(configuration.socksPort, 7_892)
    XCTAssertEqual(configuration.globalUserAgent, "mihomo-test-agent")
    XCTAssertEqual(configuration.processMatchingMode, .strict)
    XCTAssertEqual(configuration.externalResourceUserAgent.value, "mihomo-test-agent")
    XCTAssertEqual(
      configuration.externalResourceUserAgent.source,
      .mihomoConfiguration
    )
    XCTAssertEqual(configuration.tun?.isEnabled, true)
    XCTAssertEqual(configuration.tun?.stack, "system")
    XCTAssertEqual(configuration.tun?.automaticallyRoutesTraffic, true)
  }

  func testUnknownProcessMatchingModeRemainsUnavailable() throws {
    let response = try decoder.decode(
      MihomoRuntimeConfigurationResponse.self,
      from: Data("{\"find-process-mode\":\"future-mode\"}".utf8)
    )

    XCTAssertNil(response.runtimeConfiguration.processMatchingMode)
  }

  func testUsesMihomoDefaultUserAgentWhenConfigurationValueIsUnsafe() {
    let configuration = MihomoRuntimeConfiguration(
      mode: nil,
      tun: nil,
      isIPv6Enabled: nil,
      allowsLAN: nil,
      mixedPort: 7_890,
      globalUserAgent: "unsafe\r\nvalue"
    )

    XCTAssertEqual(configuration.externalResourceUserAgent, .mihomoDefault)
  }

  func testDecodesConnectionFixturesWithLeafFirstChains() throws {
    let initial = try decoder.decode(
      MihomoConnectionsSnapshot.self,
      from: FixtureLoader.data(named: "connections-initial")
    )
    let next = try decoder.decode(
      MihomoConnectionsSnapshot.self,
      from: FixtureLoader.data(named: "connections-next")
    )

    XCTAssertEqual(initial.connections.count, 2)
    XCTAssertEqual(
      initial.connections.last?.chains,
      ["Synthetic Proxy", "Synthetic Group"]
    )
    XCTAssertEqual(initial.trafficSnapshot.connections.last?.rule, "DOMAIN")
    XCTAssertGreaterThan(next.downloadTotal, initial.downloadTotal)
    XCTAssertGreaterThan(next.uploadTotal, initial.uploadTotal)
  }

  func testMapsSanitizedMetadataAndConnectionStartTime() throws {
    let response = try decoder.decode(
      MihomoConnectionsSnapshot.self,
      from: Data(
        """
        {"downloadTotal":0,"uploadTotal":0,"connections":[{
          "id":"synthetic","upload":0,"download":0,"chains":["Synthetic Proxy"],
          "start":"2026-07-30T08:00:00.123Z",
          "metadata":{"host":"synthetic-host","processPath":"/Synthetic/App"}
        }]}
        """.utf8
      )
    )

    XCTAssertEqual(
      response.trafficSnapshot.connections.first?.metadata,
      ConnectionMetadata(hostname: "synthetic-host", applicationName: "App")
    )
    XCTAssertNotNil(response.trafficSnapshot.connections.first?.startedAt)
  }

  func testRejectsEmptyAndOversizedMetadataValuesForCoverage() throws {
    let oversized = String(repeating: "x", count: 2_049)
    let data = Data(
      """
      {"downloadTotal":0,"uploadTotal":0,"connections":[{
        "id":"synthetic","upload":0,"download":0,"chains":["Synthetic Proxy"],
        "metadata":{"host":" ","sniffHost":"\(oversized)","process":"\\n","processPath":"/"}
      }]}
      """.utf8
    )

    let response = try decoder.decode(MihomoConnectionsSnapshot.self, from: data)

    XCTAssertEqual(
      response.trafficSnapshot.connections.first?.metadataAvailability,
      .unavailable
    )
  }

  func testMalformedOptionalMetadataDoesNotBreakTrafficDecoding() throws {
    let data = Data(
      """
      {"downloadTotal":1,"uploadTotal":2,"connections":[{
        "id":"synthetic","upload":2,"download":1,"chains":["Synthetic Proxy"],
        "metadata":{"host":42,"processPath":["unexpected"]}
      }]}
      """.utf8
    )

    let response = try decoder.decode(MihomoConnectionsSnapshot.self, from: data)

    XCTAssertEqual(response.trafficSnapshot.kernelTotal, TrafficBytes(upload: 2, download: 1))
    XCTAssertEqual(
      response.trafficSnapshot.connections.first?.metadataAvailability,
      .unavailable
    )
  }

  func testDoesNotUseIPAddressAsHostnameFallback() throws {
    let data = Data(
      """
      {"downloadTotal":0,"uploadTotal":0,"connections":[{
        "id":"synthetic","upload":0,"download":0,"chains":["Synthetic Proxy"],
        "metadata":{"host":"203.0.113.1","destinationIP":"203.0.113.1"}
      }]}
      """.utf8
    )

    let response = try decoder.decode(MihomoConnectionsSnapshot.self, from: data)

    XCTAssertNil(response.trafficSnapshot.connections.first?.metadata.hostname)
  }

  func testNeverPassesFullProcessPathIntoDomainMetadata() throws {
    let data = Data(
      """
      {"downloadTotal":0,"uploadTotal":0,"connections":[{
        "id":"synthetic","upload":0,"download":0,"chains":["Synthetic Proxy"],
        "metadata":{"process":"/Applications/Synthetic.app/Contents/MacOS/Synthetic"}
      }]}
      """.utf8
    )

    let response = try decoder.decode(MihomoConnectionsSnapshot.self, from: data)

    XCTAssertEqual(
      response.trafficSnapshot.connections.first?.metadata.applicationName,
      "Synthetic"
    )
  }

  func testUsesOutermostApplicationBundleFromProcessPath() throws {
    let data = Data(
      """
      {"downloadTotal":0,"uploadTotal":0,"connections":[{
        "id":"synthetic","upload":0,"download":0,"chains":["Synthetic Proxy"],
        "metadata":{
          "process":"Google Chrome Helper",
          "processPath":"/Applications/Google Chrome.app/Contents/Frameworks/Google Chrome Helper.app/Contents/MacOS/Google Chrome Helper"
        }
      }]}
      """.utf8
    )

    let response = try decoder.decode(MihomoConnectionsSnapshot.self, from: data)
    let applicationName = response.trafficSnapshot.connections.first?.metadata.applicationName

    XCTAssertEqual(applicationName, "Google Chrome")
    XCTAssertFalse(try XCTUnwrap(applicationName).contains("/"))
  }

  func testUsesApplicationBundleFromWindowsStyleProcessPath() throws {
    let data = Data(
      """
      {"downloadTotal":0,"uploadTotal":0,"connections":[{
        "id":"synthetic","upload":0,"download":0,"chains":["Synthetic Proxy"],
        "metadata":{
          "process":"com.synthetic.backend",
          "processPath":"C:\\\\Apps\\\\Synthetic.app\\\\Contents\\\\com.synthetic.backend"
        }
      }]}
      """.utf8
    )

    let response = try decoder.decode(MihomoConnectionsSnapshot.self, from: data)

    XCTAssertEqual(
      response.trafficSnapshot.connections.first?.metadata.applicationName,
      "Synthetic"
    )
  }
}

File: platform/windows/MihomoMeter.Windows.Tests/MihomoControllerModelsTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Text;
using MihomoMeter.Windows.Core.Domain;
using MihomoMeter.Windows.Core.Infrastructure.Mihomo;

namespace MihomoMeter.Windows.Tests;

[TestClass]
public sealed class MihomoControllerModelsTests
{
    [TestMethod]
    public void DecodesSharedVersionAndProxyFixtures()
    {
        var version = MihomoJsonDecoder.Decode<MihomoVersionResponse>(
            FixtureLoader.Load("version"));
        var proxies = MihomoJsonDecoder.Decode<MihomoProxiesResponse>(
            FixtureLoader.Load("proxies"));

        Assert.AreEqual("v1.19.0", version.Version);
        Assert.AreEqual("Vmess", proxies.Proxies["Synthetic Proxy"].Type);
        Assert.AreEqual("Direct", proxies.ToCatalog().TypeFor("DIRECT"));
    }

    [TestMethod]
    public void DecodesSharedConnectionFixture()
    {
        var response = MihomoJsonDecoder.Decode<MihomoConnectionsSnapshot>(
            FixtureLoader.Load("connections-next"));
        var snapshot = response.ToTrafficSnapshot();

        Assert.AreEqual(2, snapshot.Connections.Count);
        Assert.AreEqual((ulong)1_400, snapshot.KernelTotal.Upload);
        Assert.AreEqual((ulong)2_900, snapshot.KernelTotal.Download);
    }

    [TestMethod]
    public void RejectsUnsupportedPayload()
    {
        Assert.ThrowsExactly<MihomoResponseException>(() =>
            MihomoJsonDecoder.Decode<MihomoConnectionsSnapshot>("{}"u8));
        Assert.ThrowsExactly<MihomoResponseException>(() =>
            MihomoJsonDecoder.Decode<MihomoProxiesResponse>("{\"proxies\":null}"u8));
        Assert.ThrowsExactly<MihomoResponseException>(() =>
            MihomoJsonDecoder.Decode<MihomoConnectionsSnapshot>(
                """
                {
                  "downloadTotal": 1,
                  "uploadTotal": 1,
                  "connections": [
                    { "id": "duplicate", "upload": 0, "download": 0, "chains": [] },
                    { "id": "duplicate", "upload": 0, "download": 0, "chains": [] }
                  ]
                }
                """u8));
    }

    [TestMethod]
    public void DecodesOnlySanitizedMetadata()
    {
        var response = MihomoJsonDecoder.Decode<MihomoConnectionsSnapshot>(
            """
            {
              "downloadTotal": 0,
              "uploadTotal": 0,
              "connections": [{
                "id": "synthetic",
                "upload": 0,
                "download": 0,
                "chains": ["Synthetic Proxy"],
                "metadata": {
                  "host": "example.test",
                  "processPath": "C:\\Program Files\\Synthetic\\synthetic.exe"
                }
              }]
            }
            """u8);

        Assert.AreEqual(
            new ConnectionMetadata("example.test", "synthetic.exe"),
            response.ToTrafficSnapshot().Connections[0].Metadata);
    }

    [TestMethod]
    public void RejectsIpMalformedAndOversizedMetadataForCoverage()
    {
        var oversized = new string('x', 2_049);
        var payload = Encoding.UTF8.GetBytes($$"""
            {
              "downloadTotal": 0,
              "uploadTotal": 0,
              "connections": [{
                "id": "synthetic",
                "upload": 0,
                "download": 0,
                "chains": ["Synthetic Proxy"],
                "metadata": {
                  "host": "203.0.113.1",
                  "sniffHost": "{{oversized}}",
                  "process": 42,
                  "processPath": "/"
                }
              }]
            }
            """);

        var response = MihomoJsonDecoder.Decode<MihomoConnectionsSnapshot>(payload);

        Assert.AreEqual(
            ConnectionMetadata.Unavailable,
            response.ToTrafficSnapshot().Connections[0].Metadata);
    }

    [TestMethod]
    public void UsesSniffHostAndApplicationBundleFallbacks()
    {
        var response = MihomoJsonDecoder.Decode<MihomoConnectionsSnapshot>(
            """
            {
              "downloadTotal": 0,
              "uploadTotal": 0,
              "connections": [{
                "id": "synthetic",
                "upload": 0,
                "download": 0,
                "chains": ["Synthetic Proxy"],
                "metadata": {
                  "host": " ",
                  "sniffHost": "fallback.test",
                  "process": " ",
                  "processPath": "C:\\Apps\\Synthetic.app\\Contents\\helper"
                }
              }]
            }
            """u8);

        Assert.AreEqual(
            new ConnectionMetadata("fallback.test", "Synthetic"),
            response.ToTrafficSnapshot().Connections[0].Metadata);
    }

    [TestMethod]
    public void DecodesStartAndToleratesMalformedOptionalValues()
    {
        var valid = MihomoJsonDecoder.Decode<MihomoConnectionsSnapshot>(
            """
            {
              "downloadTotal": 0,
              "uploadTotal": 0,
              "connections": [{
                "id": "valid",
                "upload": 0,
                "download": 0,
                "chains": [],
                "start": "2026-08-08T01:02:03.456Z"
              }]
            }
            """u8);
        var malformed = MihomoJsonDecoder.Decode<MihomoConnectionsSnapshot>(
            """
            {
              "downloadTotal": 0,
              "uploadTotal": 0,
              "connections": [{
                "id": "malformed",
                "upload": 0,
                "download": 0,
                "chains": [],
                "start": { "future": true }
              }]
            }
            """u8);

        Assert.AreEqual(
            DateTimeOffset.Parse("2026-08-08T01:02:03.456Z"),
            valid.Connections[0].StartedAt);
        Assert.IsNull(malformed.Connections[0].StartedAt);
    }

    [DataRow("always", MihomoProcessMatchingMode.Always)]
    [DataRow(" STRICT ", MihomoProcessMatchingMode.Strict)]
    [DataRow("off", MihomoProcessMatchingMode.Off)]
    [TestMethod]
    public void MapsKnownProcessMatchingModes(
        string value,
        MihomoProcessMatchingMode expected)
    {
        var response = MihomoJsonDecoder.Decode<MihomoRuntimeConfigurationResponse>(
            Encoding.UTF8.GetBytes($$"""{"find-process-mode":"{{value}}"}"""));

        Assert.AreEqual(expected, response.ToProcessMatchingMode());
    }

    [TestMethod]
    public void KeepsUnknownProcessMatchingModeUnavailable()
    {
        var response = MihomoJsonDecoder.Decode<MihomoRuntimeConfigurationResponse>(
            """{"find-process-mode":"future"}"""u8);

        Assert.IsNull(response.ToProcessMatchingMode());
    }

    [TestMethod]
    public void DecodesRuleAndRuntimeConfiguration()
    {
        var connections = MihomoJsonDecoder.Decode<MihomoConnectionsSnapshot>(
            """
            {
              "downloadTotal": 0,
              "uploadTotal": 0,
              "connections": [{
                "id": "synthetic",
                "upload": 0,
                "download": 0,
                "chains": ["Synthetic Proxy"],
                "rule": "DOMAIN-SUFFIX"
              }]
            }
            """u8);
        var configuration = MihomoJsonDecoder.Decode<MihomoRuntimeConfigurationResponse>(
            """
            {
              "mode": "rule",
              "allow-lan": false,
              "ipv6": true,
              "mixed-port": 7890,
              "find-process-mode": "strict",
              "tun": {
                "enable": true,
                "stack": "system",
                "auto-route": true
              }
            }
            """u8).ToRuntimeConfiguration();

        Assert.AreEqual("DOMAIN-SUFFIX", connections.ToTrafficSnapshot().Connections[0].Rule);
        Assert.AreEqual("rule", configuration.Mode);
        Assert.AreEqual(true, configuration.IsTunEnabled);
        Assert.AreEqual("system", configuration.TunStack);
        Assert.AreEqual(true, configuration.AutomaticallyRoutesTraffic);
        Assert.AreEqual(true, configuration.IsIPv6Enabled);
        Assert.AreEqual(false, configuration.AllowsLan);
        Assert.AreEqual(7890, configuration.MixedPort);
        Assert.AreEqual(MihomoProcessMatchingMode.Strict, configuration.ProcessMatchingMode);
    }
}
